doc/developer/design/20231113_optimizer_notice_catalog.md

```diff
@@ -0,0 +1,547 @@
+optimization run. In order to persist those notices in
+`mz_all_optimizer_notices` we need to be able to emit retractions for a
+previously inserted notice (for example when we delete the associated catalog
+item). However, the notice structure of the notices generated by the optimizer
+is not suitable for emitting retractions - for example:
+
+1. It doesn't have an optional `created_at` field.
+2. Some of the text fields need to be resolved against catalog items that might
+   be renamed or removed in the meantime.
+
+In order to protect ourselves against invalid retraction issues arising from
+these problems we make a distinction between the notice type produced by the
+optimizer (`RawOptimizerNotice`) and the notice type stored in the
+`CatalogPlans` struct (`OptimizerNotice`).
+
+Further, the type of the notices stored in `DataflowMeatinfo::optimizer_notices`
+depends on the execution stage - within the optimizer we still need to use
+`RawOptimizerNotice`, but before storing the final result in the `CatalogPlans`
+struct we need to move to a `OptimizerNotice` vector (done with a
+`CatalogState::render_notices` call). We use the `CatalogPlans` entries to
+generate insertions and deletions in the `mz_all_optimizer_notices` table.
+
+Also, since we expect to have more than two notice types in the long term, we
+reorganize the code a bit so adding a new notice is easier in the future. The
+required steps are as follows:
+
+1. Create a `mz_transform::notice::<notice_type>` submodule.
+2. Define a struct for the new notice in that module.
+3. Implement `OptimizerNoticeApi` for that struct.
+4. Re-export the notice type in `mz_transform::notice`.
+5. Add the notice type to the `raw_optimizer_notices` macro which generates the
+   `RawOptimizerNotice` enum and other boilerplate code.
+
+#### `mz_all_optimizer_notices` maintenance
+
+Notices associated with an `object_id` are immediately removed from the
+`BuiltinTable` when the associated object is deleted. We install a
+`CatalogState::pack_optimizer_notices` helper that is
+
+- Called with `diff = 1`:
+  - In `Coordnator::bootstrap` for each re-hydrated materialized view and index.
+  - In `Coordnator::sequence_create_materialized_view` for new entries.
+  - In `Coordnator::sequence_create_index` for new entries.
+- Called with `diff = -1`:
+  - For each `id` in `drop_ids` towards the end of `Catalog::transact`.
+
+#### `mz_ignored_optimizer_notices` maintenance
+
+The design here follows the pattern adopted by MaterializeInc/materialize#21379.
+
+The `BuiltinSource` is maintained by API extensions in the `StorageController`
+in the `sequence_~` methods handling the [SQL syntax
+extensions](#sql-syntax-extensions) proposed below:
+
+- A SQL statement that marks a notice as ignored results in an insertion added
+  to the source.
+- A SQL statement that marks a notice as no longer ignored results in a
+  retraction added to the source.
+
+#### SQL syntax extensions
```

Comment:
@ParkMyCar I guess since you are currently in the SQL Council I need your approval on this as well.